In [27]:
import pandas as pd 
import numpy as np 

In [3]:
autos = pd.read_csv('autos.csv',encoding = 'Latin-1')

In [4]:
print(autos.head())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

In [5]:
info = autos.info()
print(info)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [24]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [25]:
# print colums 
col = autos.columns 
print(col)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [26]:
col_1 = col.copy()

In [9]:
print(col_1)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [70]:
def clean_col(col_1) : 
 replace = col_1.replace("yearOfRegistration","registration_year")
 replace = col_1.replace("monthOfRegistration","registration_month")
 replace = col_1.replace("notRepairedDamage","unrepaired_damage")
 replace = col_1.replace("dateCreated","ad_created")
 return replace 

autos.columns = [clean_col(c) for c in autos.columns] 
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Any columns that have mostly one value that are candidates to be dropped
Any columns that need more investigation.
Any examples of numeric data stored as text that needs to be cleaned.

In [12]:
autos.describe(include = 'all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                 ad_created  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [11]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [12]:
autos['price'].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
$1,200      639
$600        531
$3,500      498
$800        498
$2,000      460
$999        434
$750        433
$900        420
$650        419
$850        410
$700        395
$4,500      394
$300        384
$2,200      382
$950        379
$1,100      376
$1,300      371
$3,000      365
$550        356
$1,800      355
$5,500      340
$1,250      335
$350        335
$1,600      327
$1,999      322
           ... 
$10,555       1
$8,777        1
$3,075        1
$21,275       1
$3,710        1
$98,500       1
$9,590        1
$2,004        1
$2,985        1
$28,470       1
$15,870       1
$22,555       1
$198          1
$69,993       1
$5,255        1
$33,650       1
$21,700       1
$10,333       1
$5,489        1
$6,770        1
$5,740        1
$69,900       1
$4,840        1
$1,169        1
$2,874        1
$16,190       1
$14,863       1
$3,975        1
$22,299       1
$17,980       1
Name: price, Length: 235

In [13]:
autos['abtest'].value_counts()

test       25756
control    24244
Name: abtest, dtype: int64

In [14]:
autos['vehicleType'].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicleType, dtype: int64

From the analysis above two columns emerged as candidates need to be dropped. In the seller and abtest columns over 95% of the values are the same. But for the seller column one of the 50,000 sellers were private(pivat). But for the offerType column one of the 50,000 were offers(Angebot). These columns can therefore discarded due to lack of helpful information for futher analysis. these two columns need to be dropped by df.drop() method

In [15]:
autos.drop(columns = ['seller','offerType'], inplace = True)

A quick confirm two columns has been disappear

In [17]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicleType  yearOfRegistration    gearbox  powerPS  \
0  $5,000  control         bus                2004    manuell      158   
1  $8,500  control   limousine                1997  automatik      286   
2  $8,990     test   limousine                2009    manuell      102   
3  $4,350  control  kleinwagen                2007  automatik       71   
4  $1,350     test       kombi                2003    manuell        0   

    model   odometer  monthOfRegistration fuelType       brand  \
0  andere  150,000km                    3      lpg     peugeot   
1     7er  150,000km                    6   benzin         bmw   
2    golf   70,000km                    7   benzin  volkswagen   
3  fortwo   70,000km                    6   benzin       smart   
4   focus  150,000km                    7   benzin        ford   

  notRepairedDamage           ad_created  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   
3              nein  2016-03-12 00:00:00             0       33729   
4              nein  2016-04-01 00:00:00             0       39218   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

You likely found that the price and odometer columns are numeric values stored as text. For each column:
Remove any non-numeric characters using replace. 
Convert the column to a numeric dtype.
Use DataFrame.rename() to rename the column to odometer_km

In [16]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [29]:
autos['price'] = autos['price'].astype(str)
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].astype(int)
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [20]:
autos['odometer'] = autos['odometer'].astype(str)
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].astype(int)
autos['odometer']

0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 

In [18]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [19]:
autos.rename(columns = {'odometer':'odometer_km'},inplace = True)

In [22]:
# Check if the things get right 
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
price                  50000 non-null int64
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer_km            50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
ad_created             50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null object
dtypes: int64(6), object(12)
memory usage: 6.9+ MB


Now we will start analayzing on the price and odometer_km columns 

In [20]:
autos['price'].unique().shape

(2357,)

there are 2357 unique values in the price column

In [21]:
autos['odometer_km'].unique().shape

(13,)

there are 13 unique values in the odometer_km

In [25]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Please leave your comments here

In [22]:
autos['odometer_km'].describe()

count         50000
unique           13
top       150,000km
freq          32424
Name: odometer_km, dtype: object

In [27]:
autos['price'].value_counts().sort_index().head(25)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
Name: price, dtype: int64

In [23]:
autos['price'].value_counts().sort_index(ascending = False).head(25)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
Name: price, dtype: int64

In [24]:
autos['odometer_km'].value_counts().sort_index(ascending = True).head(25)

10,000km       264
100,000km     2169
125,000km     5170
150,000km    32424
20,000km       784
30,000km       789
40,000km       819
5,000km        967
50,000km      1027
60,000km      1164
70,000km      1230
80,000km      1436
90,000km      1757
Name: odometer_km, dtype: int64

Removing outliers 
df[df["col"].between(x,y)] using this method removing outliers

In [25]:
autos[autos['price'].between(1,350000)]
autos.describe()

price  yearOfRegistration       powerPS  monthOfRegistration  \
count  5.000000e+04        50000.000000  50000.000000         50000.000000   
mean   9.840044e+03         2005.073280    116.355920             5.723360   
std    4.811044e+05          105.712813    209.216627             3.711984   
min    0.000000e+00         1000.000000      0.000000             0.000000   
25%    1.100000e+03         1999.000000     70.000000             3.000000   
50%    2.950000e+03         2003.000000    105.000000             6.000000   
75%    7.200000e+03         2008.000000    150.000000             9.000000   
max    1.000000e+08         9999.000000  17700.000000            12.000000   

       nrOfPictures    postalCode  
count       50000.0  50000.000000  
mean            0.0  50813.627300  
std             0.0  25779.747957  
min             0.0   1067.000000  
25%             0.0  30451.000000  
50%             0.0  49577.000000  
75%             0.0  71540.000000  
max             0.0  99998.000000

let's move to the date columns and try to understand the date range the data covers 

In [31]:
print(autos[['dateCrawled','lastSeen','ad_created']][0:5])

           dateCrawled             lastSeen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00


In [26]:
autos['dateCrawled'].value_counts(normalize=True,dropna = False).sort_index().tail(30)

2016-04-07 10:36:19    0.00002
2016-04-07 10:36:21    0.00002
2016-04-07 10:36:24    0.00002
2016-04-07 10:36:25    0.00002
2016-04-07 10:36:35    0.00002
2016-04-07 10:36:36    0.00002
2016-04-07 10:36:37    0.00004
2016-04-07 11:06:33    0.00002
2016-04-07 11:36:19    0.00002
2016-04-07 11:36:23    0.00002
2016-04-07 11:36:24    0.00002
2016-04-07 11:36:25    0.00002
2016-04-07 11:36:34    0.00004
2016-04-07 11:36:35    0.00002
2016-04-07 12:06:19    0.00002
2016-04-07 12:06:23    0.00002
2016-04-07 12:25:34    0.00002
2016-04-07 12:25:35    0.00002
2016-04-07 13:06:18    0.00002
2016-04-07 13:25:39    0.00002
2016-04-07 13:36:19    0.00002
2016-04-07 13:36:20    0.00002
2016-04-07 13:36:37    0.00002
2016-04-07 13:36:38    0.00002
2016-04-07 14:07:04    0.00002
2016-04-07 14:30:09    0.00002
2016-04-07 14:30:26    0.00002
2016-04-07 14:36:44    0.00002
2016-04-07 14:36:55    0.00002
2016-04-07 14:36:56    0.00002
Name: dateCrawled, dtype: float64

# Write a markdown cell exploration to explain my obseravations
it looks as if the date was crawled everyday from the early march to early april, the values seem to be not distributed enough. it ran around from 0.00002 to the highest value 0.00004

In [49]:
 autos['ad_created'].value_counts(normalize= True, dropna = False).sort_index()

2015-06-11 00:00:00    0.00002
2015-08-10 00:00:00    0.00002
2015-09-09 00:00:00    0.00002
2015-11-10 00:00:00    0.00002
2015-12-05 00:00:00    0.00002
2015-12-30 00:00:00    0.00002
2016-01-03 00:00:00    0.00002
2016-01-07 00:00:00    0.00002
2016-01-10 00:00:00    0.00004
2016-01-13 00:00:00    0.00002
2016-01-14 00:00:00    0.00002
2016-01-16 00:00:00    0.00002
2016-01-22 00:00:00    0.00002
2016-01-27 00:00:00    0.00006
2016-01-29 00:00:00    0.00002
2016-02-01 00:00:00    0.00002
2016-02-02 00:00:00    0.00004
2016-02-05 00:00:00    0.00004
2016-02-07 00:00:00    0.00002
2016-02-08 00:00:00    0.00002
2016-02-09 00:00:00    0.00004
2016-02-11 00:00:00    0.00002
2016-02-12 00:00:00    0.00006
2016-02-14 00:00:00    0.00004
2016-02-16 00:00:00    0.00002
2016-02-17 00:00:00    0.00002
2016-02-18 00:00:00    0.00004
2016-02-19 00:00:00    0.00006
2016-02-20 00:00:00    0.00004
2016-02-21 00:00:00    0.00006
                        ...   
2016-03-09 00:00:00    0.03324
2016-03-

There is a wide range of ad created dates. Most fall within a month or 2 months of the listing date, but a few of them are quite old

In [50]:
autos['lastSeen'].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
2016-03-05 15:28:10    0.00002
2016-03-05 15:41:30    0.00002
2016-03-05 15:45:43    0.00002
2016-03-05 15:47:38    0.00002
2016-03-05 15:47:44    0.00002
2016-03-05 16:45:57    0.00002
2016-03-05 16:47:28    0.00002
2016-03-05 17:15:45    0.00002
2016-03-05 17:16:12    0.00002
2016-03-05 17:16:14    0.00002
2016-03-05 17:16:23    0.00002
2016-03-05 17:17:02    0.00002
2016-03-05 17:39:19    0.00002
2016-03-05 17:40:14    0.00002
2016-03-05 17:44:50    0.00002
2016-03-05 17:44:54    0.00002
2016-03-05 17:46:01    0.00002
2016-03-05 18:17:58    0.00002
2016-03-05 18:47:14    0.00002
2016-03-05 18:50:38    0.00002
2016-03-05 19:15:08    0.00002
2016-03-05 19:15:20    0.00002
2016-03-05 19:15:42    0.00002
2016-03-05 19:16:36    0.00002
2016-03-05 19:17:17    0.00002
                        ...   
2016-04-07 14:58:09    0.00004
2016-04-

# Using Series.describe() to understand the distribution of registration_year

In [27]:
print(autos['yearOfRegistration'].describe())

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: yearOfRegistration, dtype: float64


# from those statistics above We are explore insights the yearofregistration
there are 50000 values, the maximun number of values nearly 10000 and the minimun number of the column is 1000 and so on other statistics and you can explore it by yourself such as 25% or std. 

In [28]:
print(autos['yearOfRegistration'].value_counts().sort_index())

1000       1
1001       1
1111       1
1500       1
1800       2
1910       9
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       5
1957       2
1958       4
1959       7
1960      34
1961       6
1962       4
1963       9
        ... 
2001    2703
2002    2533
2003    2727
2004    2737
2005    3015
2006    2708
2007    2304
2008    2231
2009    2098
2010    1597
2011    1634
2012    1323
2013     806
2014     666
2015     399
2016    1316
2017    1453
2018     492
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: yearOfRegistration, Length: 97, dtype: int64


In [4]:
print(autos['yearOfRegistration'].describe())

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: yearOfRegistration, dtype: float64


In [5]:
print(autos['yearOfRegistration'].value_counts().sort_index())

1000       1
1001       1
1111       1
1500       1
1800       2
1910       9
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       5
1957       2
1958       4
1959       7
1960      34
1961       6
1962       4
1963       9
        ... 
2001    2703
2002    2533
2003    2727
2004    2737
2005    3015
2006    2708
2007    2304
2008    2231
2009    2098
2010    1597
2011    1634
2012    1323
2013     806
2014     666
2015     399
2016    1316
2017    1453
2018     492
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: yearOfRegistration, Length: 97, dtype: int64


the crrowst with just only 3000 registraegistrations  registered is month zero with over 5 thousands registration
the least registered is february with just only 3000 plus 8 registrations  

In [49]:
autos[autos['yearOfRegistration'] ==1000]

dateCrawled                                   name  price  \
22316  2016-03-29 16:56:41  VW_Kaefer.__Zwei_zum_Preis_von_einem.   1500   

        abtest vehicleType  yearOfRegistration  gearbox  powerPS   model  \
22316  control         NaN                1000  manuell        0  kaefer   

      odometer_km  monthOfRegistration fuelType       brand notRepairedDamage  \
22316     5,000km                    0   benzin  volkswagen               NaN   

                ad_created  nrOfPictures  postalCode             lastSeen  
22316  2016-03-29 00:00:00             0       48324  2016-03-31 10:15:28

In [52]:
autos[autos['yearOfRegistration'] == 9999]

dateCrawled                                               name  \
8012   2016-03-23 16:43:29                      Opel_GT_Karosserie_mit_Brief!   
14341  2016-03-23 01:36:20                                  Hole_kostenlos_ab   
33950  2016-03-23 21:52:25  58er_karmann_ghia_lowlight_Kaefer__zum_restaur...   
38076  2016-04-04 22:54:47                                 Mercedes_Benz_A180   

       price abtest vehicleType  yearOfRegistration gearbox  powerPS  \
8012     700   test         NaN                9999     NaN        0   
14341      0   test         NaN                9999     NaN        0   
33950   7999   test         NaN                9999     NaN        0   
38076  18000   test         NaN                9999     NaN        0   

          model odometer_km  monthOfRegistration fuelType          brand  \
8012     andere    10,000km                    0      NaN           opel   
14341       NaN    10,000km                    0      NaN            bmw   
33950    kaefer    10,000km                    0      NaN     volkswagen   
38076  a_klasse    10,000km                    0   benzin  mercedes_benz   

      notRepairedDamage           ad_created  nrOfPictures  postalCode  \
8012                NaN  2016-03-23 00:00:00             0       21769   
14341               NaN  2016-03-23 00:00:00             0       32689   
33950               NaN  2016-03-23 00:00:00             0       47638   
38076               NaN  2016-04-04 00:00:00             0       51379   

                  lastSeen  
8012   2016-04-05 20:16:15  
14341  2016-03-23 08:47:00  
33950  2016-04-06 03:46:40  
38076  2016-04-07 02:44:52

With max and min method, I decided that the highest is 9999 and the lowest is 1000 

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic

In [4]:
autos[(autos['yearOfRegistration'] < 1900)
      |(autos['yearOfRegistration'] > 2016)]['yearOfRegistration'].count() 

1972

In [8]:
autos[(autos['yearOfRegistration'] < 1900)
      |(autos['yearOfRegistration'] > 2016)]['yearOfRegistration'].value_counts()

2017    1453
2018     492
9999       4
5000       4
2019       3
9000       2
1800       2
6200       1
4500       1
8888       1
4800       1
2800       1
1001       1
1000       1
1111       1
1500       1
9996       1
5911       1
4100       1
Name: yearOfRegistration, dtype: int64

there are 1917 the listings fall outside 1900-2016 range(about 5% all of the listings)
and most of them occur in 2017,2018 as I showed above. 

In [10]:
autos[(autos['yearOfRegistration']>1899)
      &(autos['yearOfRegistration']<2000)]['yearOfRegistration'].value_counts().sort_index()

1910       9
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       5
1957       2
1958       4
1959       7
1960      34
1961       6
1962       4
1963       9
1964      12
1965      17
1966      22
1967      27
1968      26
        ... 
1970      45
1971      27
1972      35
1973      26
1974      24
1975      19
1976      27
1977      22
1978      47
1979      35
1980      97
1981      31
1982      43
1983      53
1984      53
1985     105
1986      76
1987      75
1988     142
1989     181
1990     395
1991     356
1992     391
1993     445
1994     660
1995    1313
1996    1444
1997    2028
1998    2453
1999    3000
Name: yearOfRegistration, Length: 61, dtype: int64

Now remove the values those upper and lower bounds and calculatate the distribution of the remaning values using the Series.values_counts(normalize = True)

In [12]:
autos.shape

(50000, 20)

In [29]:
autos[autos['monthOfRegistration'] == 0 ].count()

dateCrawled            5075
name                   5075
seller                 5075
offerType              5075
price                  5075
abtest                 5075
vehicleType            3123
yearOfRegistration     5075
gearbox                3764
powerPS                5075
model                  4173
odometer               5075
monthOfRegistration    5075
fuelType               3224
brand                  5075
notRepairedDamage      2031
dateCreated            5075
nrOfPictures           5075
postalCode             5075
lastSeen               5075
dtype: int64

In [32]:
autos = autos[autos['yearOfRegistration'].between(1910,2016)]
autos.shape

(48028, 20)

So 1972 car listings were removed. Now we will also remove values which have a registration_month of 0. As was stated earlier there were 5057 rows with a registration_month of 0, however, some of the listings we removed from outside our acceptable registration_year range also had a registration month of 0 so now we will find out how many car listings with a registration month of 0 were within our acceptable registration_year range.

In [36]:
autos = autos[autos['monthOfRegistration'] !=0] 
autos.shape

(43441, 20)

there are 48028 values not equal to zero in the monthofregistration column. 
About 4587  listings were removed, which represents the cars that were within our our registration_year range but did not match our registration_month criteria. Now we will calculate the distribution of the remaining registration_month values using Series.value_counts(normalize=True).

In [44]:
autos['yearOfRegistration'].value_counts(normalize = True).sort_index().head(100)
                                                                        

1910    0.000023
1927    0.000023
1929    0.000023
1931    0.000023
1934    0.000023
1937    0.000069
1938    0.000023
1939    0.000023
1941    0.000046
1943    0.000023
1948    0.000023
1950    0.000046
1951    0.000046
1952    0.000023
1953    0.000023
1954    0.000023
1955    0.000046
1956    0.000092
1957    0.000023
1958    0.000046
1959    0.000161
1960    0.000368
1961    0.000138
1962    0.000092
1963    0.000161
1964    0.000253
1965    0.000368
1966    0.000460
1967    0.000529
1968    0.000529
          ...   
1987    0.001519
1988    0.002900
1989    0.003522
1990    0.005962
1991    0.007136
1992    0.007689
1993    0.008310
1994    0.012868
1995    0.023411
1996    0.027716
1997    0.040469
1998    0.049124
1999    0.061969
2000    0.061670
2001    0.056629
2002    0.054281
2003    0.059023
2004    0.060151
2005    0.062867
2006    0.059920
2007    0.051012
2008    0.049492
2009    0.046960
2010    0.036095
2011    0.036878
2012    0.030041
2013    0.018301
2014    0.0151

it seems that the majority of vehicles were first registered within the past 20 years(from 2000 till 2020) 

in the next part we are going to work with the brand column to extensive understand of your data.

In [58]:
autos['brand'].value_counts(normalize = True).sort_index()

alfa_romeo        0.006561
audi              0.087981
bmw               0.112198
chevrolet         0.005709
chrysler          0.003637
citroen           0.013812
dacia             0.002716
daewoo            0.001496
daihatsu          0.002394
fiat              0.025483
ford              0.068484
honda             0.007850
hyundai           0.010129
jaguar            0.001611
jeep              0.002325
kia               0.007550
lada              0.000552
lancia            0.001013
land_rover        0.002141
mazda             0.014940
mercedes_benz     0.099192
mini              0.009254
mitsubishi        0.007942
nissan            0.015377
opel              0.104464
peugeot           0.029811
porsche           0.006376
renault           0.046845
rover             0.001312
saab              0.001680
seat              0.018324
skoda             0.016804
smart             0.014272
sonstige_autos    0.009599
subaru            0.002049
suzuki            0.006008
toyota            0.013190
t

My selection 
I need to choose that have over a certain percentage of the total values 5% 

In [8]:
brand_counts = autos['brand'].value_counts(normalize = True).sort_index()
popular_brands = brand_counts[brand_counts > 0.05].index
popular_brands

Index(['audi', 'bmw', 'ford', 'mercedes_benz', 'opel', 'volkswagen'], dtype='object')

there are 6 popular brands that fit this criteria. let's filter out our dataframe to only consist of these popular brands and store the filtered result in a new dataframe 

In [ ]:
Now we will look at the mean prices per brand

In [ ]:
# create an empty dictionary to hold all aggregate data 
mean_brand = {}
# using Loop over your selected brands, and assign the mean price to the dictionary, with the brand name as the key
for brand in popular_brands : 
    brand_name = autos[autos["brand"] == brand]
    avg_brand_name = brand_name["price"].mean()
    mean_by_brand[price] = int(avg_brand_name)
    
mean_brand